pyomo/repn/plugins/gams_writer.py

```diff
@@ -92,6 +92,21 @@ def visiting_potential_leaf(self, node):
         if node.__class__ in native_types:
             return True, str(node)
 
+        # Make sure all components in active constraints are basic ctypes we
+        # know how to deal with. This means anything with a type() method must
+        # be one of the following allowable types.
+        if hasattr(node, "type"):
```

Comment:
Where you say "At this point, this had better be a Pyomo type", we can still get `NumericConstant` types here, and those do not have a `type()` method. Is this contrary to what you expected? And what do you then think should be done about that?

Reply:
Hmm - that is contrary to what I expected.  I think I would be in favor of adding `type()` to `NumericConstant`.  Thoughts, @whart222?

Reply:
I think that adding type() would be confusing, since this is a component type ... and NumericConstants are not components.  I think it's sensible to check for is_expression_type() as John suggests, and avoid calling this function.  (More generally, I've seen several posts about Python that suggest that developers avoid the use of hasattr(), since it's behavior is not as reliable as you might guess.)

Reply:
So currently, the following code will not work:

```python
def visiting_potential_leaf(self, node):
    """
    Visiting a potential leaf.

    Return True if the node is not expanded.
    """
    if node is None:
        return True, None

    if node.__class__ in native_types:
        return True, str(node)

    if node.is_expression_type():
        # we will descend into this, so type checking will happen later
        return False, None

    # At this point, this should be a Pyomo type because native types and 
    # expressios have already been filtered, so make sure all components in
    # active constraints are basic ctypes we know how to deal with.
    if self.ctype(node) not in (Var, Param, Expression, Objective):
        raise RuntimeError(
            "Unallowable component '%s' of type %s found in an active "
            "constraint or objective.\nThe GAMS writer cannot export "
            "expressions with this component type."
            % (node.name, self.ctype(node).__name__))
    if self.ctype(node) in (Param, Expression, Objective):
        # For these, make sure it's on the right model. We can check
        # Vars later since they don't disappear from the expressions
        self.treechecker.on_tree(node)

    if node.is_variable_type():
        if node.fixed:
            return True, str(value(node))
        label = self.smap.getSymbol(node)
        return True, label

    return True, str(value(node))

def ctype(self, comp):
    if isinstance(comp, IComponent):
        return comp.ctype
    else:
        return comp.type()
```

Are you saying we should just change the `ctype` helper function to do this, and also allow `NumericConstant` in the list above?

```python
def ctype(self, comp):
    if isinstance(comp, IComponent):
        return comp.ctype
    elif isinstance(comp, Component):
        return comp.type()
    else:
        return type(comp)
```